From 8dcaef5d114ae7416fd292996b9a80244f42e39e Mon Sep 17 00:00:00 2001
From: Sarosh Hasan <quic_sarohasa@quicinc.com>
Date: Fri, 26 Jul 2024 20:08:50 +0530
Subject: [PATCH 14/14] net: stmmac: dwmac-qcom-ethqos: Enable SCMI ETH

On sa8255p platforms, resource such as clocks, regulators etc.
are managed in the firmware. The Linux driver uses power and
performance domains to request resource operations. The domains
are turned on/off using runtime PM framework.

Power domains represents suppliers to put device into various
power state. Two modeled power states d1 and d3 are supported.
There are two power domain suppliers named "core" and "mdio".

D0 <--> D3 :: Turn on/off both suppliers
D0 <--> D1 :: Turn on/off 'mdio' supplier

Performance domain represents the serdes phy with link speed
being used as performance levels.

Change-Id: I5bc626ef7abccdd4311b9a2afdbe880592bfd705
Signed-off-by: Sarosh Hasan <quic_sarohasa@quicinc.com>
---
 .../stmicro/stmmac/dwmac-qcom-ethqos.c        | 175 +++++++++++++++++-
 1 file changed, 168 insertions(+), 7 deletions(-)
 
diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c b/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c
index 749c91b7d02d..14f81dea0fe7 100644
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c
@@ -7,6 +7,8 @@
 #include <linux/platform_device.h>
 #include <linux/phy.h>
 #include <linux/phy/phy.h>
+#include <linux/pm_opp.h>
+#include <linux/pm_domain.h>
 
 #include "stmmac.h"
 #include "stmmac_platform.h"
@@ -85,6 +87,12 @@
 
 #define SGMII_10M_RX_CLK_DVDR			0x31
 
+enum domain_t {
+	POWER_CORE = 0,
+	POWER_MDIO = 1,
+	PERF_SERDES = 2,
+};
+
 struct ethqos_emac_por {
 	unsigned int offset;
 	unsigned int value;
@@ -98,6 +106,7 @@ struct ethqos_emac_driver_data {
 	const char *link_clk_name;
 	bool has_integrated_pcs;
 	u32 dma_addr_width;
+	unsigned int ptp_clk_rate;
 	struct dwmac4_addrs dwmac4_addrs;
 	bool needs_sgmii_loopback;
 };
@@ -120,6 +129,7 @@ struct qcom_ethqos {
 	bool rgmii_config_loopback_en;
 	bool has_emac_ge_3;
 	bool needs_sgmii_loopback;
+	struct dev_pm_domain_list *pd_list;
 };
 
 static int rgmii_readl(struct qcom_ethqos *ethqos, unsigned int offset)
@@ -299,6 +309,7 @@ static const struct ethqos_emac_driver_data emac_v4_0_0_data = {
 	.has_integrated_pcs = true,
 	.needs_sgmii_loopback = true,
 	.dma_addr_width = 36,
+	.ptp_clk_rate = 230400000,
 	.dwmac4_addrs = {
 		.dma_chan = 0x00008100,
 		.dma_chan_offset = 0x1000,
@@ -317,6 +328,52 @@ static const struct ethqos_emac_driver_data emac_v4_0_0_data = {
 	},
 };
 
+static int qcom_ethqos_domain_attach(struct qcom_ethqos *ethqos)
+{
+	struct dev_pm_domain_attach_data pd_data = {
+		.pd_flags = PD_FLAG_NO_DEV_LINK,
+		.pd_names = (const char*[]) { "power_core", "power_mdio", "perf_serdes" },
+		.num_pd_names = 3,
+	};
+	struct device *dev = &ethqos->pdev->dev;
+
+	return dev_pm_domain_attach_list(dev, &pd_data, &ethqos->pd_list);
+}
+
+static int qcom_ethqos_domain_on(struct qcom_ethqos *ethqos, enum domain_t dom)
+{
+	struct device *dev = ethqos->pd_list->pd_devs[dom];
+	int ret = 0;
+
+	ret = pm_runtime_resume_and_get(dev);
+	if (ret < 0)
+		dev_err(dev, "poweron(domain=%d) failed.(err=%d)\n", dom, ret);
+
+	return ret;
+}
+
+static void qcom_ethqos_domain_off(struct qcom_ethqos *ethqos, enum domain_t dom)
+{
+	struct device *dev = ethqos->pd_list->pd_devs[dom];
+	int ret = 0;
+
+	ret = pm_runtime_put_sync(dev);
+	if (ret < 0)
+		dev_err(dev, "poweroff(domain=%d) failed.(err=%d)\n", dom, ret);
+}
+
+static int qcom_ethqos_serdes_set_level(struct qcom_ethqos *ethqos)
+{
+	struct device *dev = ethqos->pd_list->pd_devs[PERF_SERDES];
+	int ret = 0;
+
+	ret = dev_pm_opp_set_level(dev, ethqos->speed);
+	if (ret)
+		dev_err(dev, "Failed to set serdes level. (err=%d)\n", ret);
+
+	return ret;
+}
+
 static int ethqos_dll_configure(struct qcom_ethqos *ethqos)
 {
 	struct device *dev = &ethqos->pdev->dev;
@@ -733,6 +790,86 @@ static void qcom_ethqos_serdes_powerdown(struct net_device *ndev, void *priv)
 	phy_exit(ethqos->serdes_phy);
 }
 
+static int qcom_ethqos_serdes_up(struct net_device *ndev, void *priv)
+{
+	struct qcom_ethqos *ethqos = priv;
+	int ret = 0;
+
+	ret = qcom_ethqos_serdes_set_level(ethqos);
+
+	return ret;
+}
+
+static void qcom_ethqos_serdes_down(struct net_device *ndev, void *priv)
+{
+	struct qcom_ethqos *ethqos = priv;
+
+	qcom_ethqos_domain_off(ethqos, PERF_SERDES);
+}
+
+/* callback for stmmac runtime suspend/resume functions */
+static int qcom_ethqos_domain_transition_d0d1(void *priv, bool high)
+{
+	struct qcom_ethqos *ethqos = priv;
+	int ret = 0;
+
+	if (high) {
+		ret = qcom_ethqos_domain_on(ethqos, POWER_MDIO);
+		if (ret < 0) {
+			dev_err(&ethqos->pdev->dev, "Transition from d1 to d0 failed\n");
+			return ret;
+		}
+		ethqos_set_func_clk_en(ethqos);
+	} else {
+		qcom_ethqos_domain_off(ethqos, POWER_MDIO);
+	}
+
+	return ret;
+}
+
+static int qcom_ethqos_domain_transition_d0d3(void *priv, bool high)
+{
+	struct qcom_ethqos *ethqos = priv;
+	int ret = 0;
+
+	if (high) {
+		ret = qcom_ethqos_domain_on(ethqos, POWER_CORE);
+		if (ret < 0) {
+			dev_err(&ethqos->pdev->dev, "CORE Transition from d3 to d0 failed\n");
+			return ret;
+		}
+
+		ret = qcom_ethqos_domain_on(ethqos, POWER_MDIO);
+		if (ret < 0) {
+			dev_err(&ethqos->pdev->dev, "MDIO Transition from d3 to d0 failed\n");
+			qcom_ethqos_domain_off(ethqos, POWER_CORE);
+			return ret;
+		}
+
+	} else {
+		qcom_ethqos_domain_off(ethqos, POWER_MDIO);
+		qcom_ethqos_domain_off(ethqos, POWER_CORE);
+	}
+
+	return ret;
+}
+
+static void qcom_ethqos_exit(struct platform_device *pdev, void *prv)
+{
+	struct net_device *ndev = platform_get_drvdata(pdev);
+	struct stmmac_priv *priv = netdev_priv(ndev);
+	struct qcom_ethqos *ethqos = prv;
+
+	/* pm runtime is disabled on driver remove.
+	 */
+	if (!pm_runtime_enabled(priv->device)) {
+		pm_runtime_enable(priv->device);
+		qcom_ethqos_domain_transition_d0d3(prv, false);
+		pm_runtime_disable(priv->device);
+	}
+	dev_pm_domain_detach_list(ethqos->pd_list);
+}
+
 static int ethqos_clks_config(void *priv, bool enabled)
 {
 	struct qcom_ethqos *ethqos = priv;
@@ -788,7 +925,7 @@ static int qcom_ethqos_probe(struct platform_device *pdev)
 	struct stmmac_resources stmmac_res;
 	struct device *dev = &pdev->dev;
 	struct qcom_ethqos *ethqos;
-	int ret, i;
+	int ret, i, use_domains = 0;
 
 	ret = stmmac_get_platform_resources(pdev, &stmmac_res);
 	if (ret)
@@ -830,6 +967,25 @@ static int qcom_ethqos_probe(struct platform_device *pdev)
 	}
 
 	ethqos->pdev = pdev;
+
+	if (of_device_is_compatible(np, "qcom,sa8255p-ethqos")) {
+		ret = qcom_ethqos_domain_attach(ethqos);
+		if (ret < 0) {
+			dev_err(dev, "Failed to attach domains.\n");
+			return ret;
+		}
+
+		dev_info(dev, "domains loaded successfully.\n");
+		use_domains = 1;
+		plat_dat->clks_config = qcom_ethqos_domain_transition_d0d1;
+		plat_dat->serdes_powerup = qcom_ethqos_serdes_up;
+		plat_dat->serdes_powerdown = qcom_ethqos_serdes_down;
+		plat_dat->exit = qcom_ethqos_exit;
+
+		qcom_ethqos_domain_transition_d0d3(ethqos, true);
+		qcom_ethqos_domain_on(ethqos, PERF_SERDES);
+	}
+
 	ethqos->rgmii_base = devm_platform_ioremap_resource_byname(pdev, "rgmii");
 	if (IS_ERR(ethqos->rgmii_base))
 		return dev_err_probe(dev, PTR_ERR(ethqos->rgmii_base),
@@ -844,11 +1000,12 @@ static int qcom_ethqos_probe(struct platform_device *pdev)
 	ethqos->has_emac_ge_3 = data->has_emac_ge_3;
 	ethqos->needs_sgmii_loopback = data->needs_sgmii_loopback;
 
-	ethqos->link_clk = devm_clk_get(dev, data->link_clk_name ?: "rgmii");
-	if (IS_ERR(ethqos->link_clk))
-		return dev_err_probe(dev, PTR_ERR(ethqos->link_clk),
-				     "Failed to get link_clk\n");
-
+	if (!use_domains) {
+		ethqos->link_clk = devm_clk_get(dev, data->link_clk_name ?: "rgmii");
+		if (IS_ERR(ethqos->link_clk))
+			return dev_err_probe(dev, PTR_ERR(ethqos->link_clk),
+						 "Failed to get link_clk\n");
+	}
 	ret = ethqos_clks_config(ethqos, true);
 	if (ret)
 		return ret;
@@ -870,7 +1027,10 @@ static int qcom_ethqos_probe(struct platform_device *pdev)
 	plat_dat->bsp_priv = ethqos;
 	plat_dat->fix_mac_speed = ethqos_fix_mac_speed;
 	plat_dat->dump_debug_regs = rgmii_dump;
-	plat_dat->ptp_clk_freq_config = ethqos_ptp_clk_freq_config;
+	if (use_domains)
+		plat_dat->clk_ptp_rate = data->ptp_clk_rate;
+	else
+		plat_dat->ptp_clk_freq_config = ethqos_ptp_clk_freq_config;
 	plat_dat->has_gmac4 = 1;
 	if (ethqos->has_emac_ge_3)
 		plat_dat->dwmac4_addrs = &data->dwmac4_addrs;
@@ -899,6 +1059,7 @@ static int qcom_ethqos_probe(struct platform_device *pdev)
 static const struct of_device_id qcom_ethqos_match[] = {
 	{ .compatible = "qcom,qcs404-ethqos", .data = &emac_v2_3_0_data},
 	{ .compatible = "qcom,qcs8300-ethqos", .data = &emac_v4_0_0_data},
+	{ .compatible = "qcom,sa8255p-ethqos", .data = &emac_v4_0_0_data},
 	{ .compatible = "qcom,sa8775p-ethqos", .data = &emac_v4_0_0_data},
 	{ .compatible = "qcom,sc8280xp-ethqos", .data = &emac_v3_0_0_data},
 	{ .compatible = "qcom,sm8150-ethqos", .data = &emac_v2_1_0_data},
